File: backend/src/main/java/com/mhbilliards/billiards_management/service/tableBilliardType/TableBilliardTypeService.java
package com.mhbilliards.billiards_management.service.tableBilliardType;

import java.util.List;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import com.mhbilliards.billiards_management.dto.tableBilliardType.TableBilliardTypeRequest;
import com.mhbilliards.billiards_management.dto.tableBilliardType.TableBilliardTypeResponse;

public interface TableBilliardTypeService {
    TableBilliardTypeResponse createTableBilliardType(TableBilliardTypeRequest request);

    TableBilliardTypeResponse updateTableBilliardType(Long id, TableBilliardTypeRequest request);

    TableBilliardTypeResponse getTableBilliardTypeById(Long id);

    Page<TableBilliardTypeResponse> getPageTableBilliardTypes(Pageable pageable);

    void deleteTableBilliardType(Long id);

    List<TableBilliardTypeResponse> getAllTableBilliardTypes();
}

File: frontend/src/services/tableBilliardTypeService.ts
import { ApiResponse, PageResponse, PaginationParams } from "@/types/base";
import {
    ITableBilliardTypeRequest,
    ITableBilliardTypeResponse,
} from "@/types/tableBilliardType";
import axiosClient from "./axiosClient";

const API_URL = "/table-billiard-types";

const tableBilliardTypeService = {
    createTableBilliardType: async (
        req: ITableBilliardTypeRequest,
    ): Promise<ApiResponse<ITableBilliardTypeResponse>> => {
        const res = await axiosClient.post(API_URL, req);
        return res.data;
    },

    updateTableBilliardType: async (
        id: number,
        req: ITableBilliardTypeRequest,
    ): Promise<ApiResponse<ITableBilliardTypeResponse>> => {
        const res = await axiosClient.put(`${API_URL}/${id}`, req);
        return res.data;
    },

    getTableBilliardTypeById: async (
        id: number,
    ): Promise<ApiResponse<ITableBilliardTypeResponse>> => {
        const res = await axiosClient.get(`${API_URL}/${id}`);
        return res.data;
    },

    getPageTableBilliardTypes: async (
        params: PaginationParams,
    ): Promise<ApiResponse<PageResponse<ITableBilliardTypeResponse>>> => {
        const res = await axiosClient.get(API_URL, {
            params: {
                page: params.page ?? 0,
                size: params.size ?? 10,
                sortBy: params.sortBy ?? "createdAt",
                sortDirection: params.sortDirection ?? "desc",
            },
        });
        return res.data;
    },

    getAllTableBilliardTypes: async (): Promise<
        ApiResponse<ITableBilliardTypeResponse[]>
    > => {
        const res = await axiosClient.get(`${API_URL}/all`);
        return res.data;
    },

    deleteTableBilliardType: async (id: number): Promise<ApiResponse<null>> => {
        const res = await axiosClient.delete(`${API_URL}/${id}`);
        return res.data;
    },
};

export default tableBilliardTypeService;
